src/frontend/src/optimizer/plan_node/logical_filter.rs

```diff
@@ -175,6 +175,14 @@ impl ColPrunable for LogicalFilter {
     }
 }
 
+impl ExprRewritable for LogicalFilter {
+    fn rewrite_exprs(&self, r: &mut dyn ExprRewriter) -> PlanRef {
+        let mut new = self.clone();
+        new.core.rewrite_exprs(r);
+        new.into()
+    }
+}
+
```

Comment:
I have some concerns about the `PlanId` of `PlanRef`. Since we are trying to mutate the `LogicalFilter`, the original filter and the mutated filter should have different `PlanId` unless it is inevitable.

Reply:
Meanwhile I think directly calling method `clone` of any `PlanNode` is not a best practice, that is also why method `clone_with_inputs` exists. Even though we just have a rule based heuristic optimizer which only rewrites a single plan tree currently, both `clone` and `clone_with_inputs` work well, however, if we want to generate multiple candidate plans and choose the best one in the future, the `PlanId` matters.

Reply:
Hmm, indeed. `PlanId` is a cheap resource anyway, and if we are going to clone all the data, we may as well clone it with a new planid.

Reply:
> Meanwhile I think directly calling method clone of any PlanNode is not a best practice

However, it is ok if we are simply cloning the plannode without any modification to it, correct?

Reply:
> However, it is ok if we are simply cloning the plannode without any modification to it or its inputs, correct?

Simply `clone` the `PlanNode` will result in two different allocation of the same `PlanNode`, however, our optimizer mostly considers whether two `PlanNode`s are equal by `PlanNodeId` or `PlanRef`(Two `Rc`s are equal if their inner values are equal, even if they are stored in different allocation), so I think it is generally acceptable.

Reply:
I suddenly realize that two different instance of share nodes with the same PlanNodeId will make DAG plan tricky. Considering predicate push down of share node, it only push down the predicate until all its parents are visited, but we can only replace one of the two different instance of share nodes input. So I think it should be no more than one plan node with the same PlanNodeId exists, at least for share node.

Reply:
@chenzl25 I decided to never clone nodes that do not need to rewrite exprs. 

Hence, LogicalShare will not be cloned from https://github.com/risingwavelabs/risingwave/pull/7541 onwards